src/lib/optimizer/strategy/subquery_to_join_rule.cpp

```diff
@@ -0,0 +1,564 @@
+    info.right_operand = left_side;
+  } else {
+    return std::nullopt;
+  }
+
+  if (!predicate_node->find_column_id(*info.right_operand)) {
+    return std::nullopt;
+  }
+
+  auto expression_it = parameter_mapping.find(parameter_id);
+  if (expression_it == parameter_mapping.end()) {
+    return std::nullopt;
+  }
+
+  info.left_operand = expression_it->second;
+  return info;
+}
+
+/**
+ * Finds predicate nodes to pull up to become join predicates and other nodes which need to be removed or patched.
+ *
+ * This selects all predicates that can be safely pulled up and turned into join predicates. It also collects all
+ * projection nodes above the selected predicates. These need to be removed to make sure that the required columns are
+ * available for the pulled up predicates. Finally, it collects aggregate nodes which need to be patched with
+ * additional group by columns to allow the predicates to be pulled up.
+ */
+PredicatePullUpInfo prepare_predicate_pull_up(
+    const std::shared_ptr<AbstractLQPNode>& lqp,
+    const std::map<ParameterID, std::shared_ptr<AbstractExpression>>& parameter_mapping) {
+  // We are only interested in predicate, projection, validate, aggregate and sort nodes. These only ever have one
+  // input, thus we can scan the path of nodes linearly. This makes it easy to track which projection nodes actually
+  // need to be removed, and which predicates need to be considered when patching aggregates.
+  PredicatePullUpInfo info{};
+  size_t num_projections_to_remove = 0;
+  size_t num_aggregates_to_patch = 0;
+
+  auto node = lqp;
+  while (node != nullptr) {
+    if (node->type == LQPNodeType::Projection) {
+      info.projection_nodes.emplace_back(std::static_pointer_cast<ProjectionNode>(node));
+    } else if (node->type == LQPNodeType::Predicate) {
+      const auto& predicate_node = std::static_pointer_cast<PredicateNode>(node);
+      auto maybe_predicate_info =
+          should_become_join_predicate(predicate_node, parameter_mapping, !info.aggregate_nodes.empty());
+      if (maybe_predicate_info) {
+        info.predicates.emplace_back(*maybe_predicate_info);
+        info.predicate_nodes.emplace(predicate_node);
+        // All projections/aggregates found so far need to be removed/patched
+        num_projections_to_remove = info.projection_nodes.size();
+        num_aggregates_to_patch = info.aggregate_nodes.size();
+      }
+    } else if (node->type == LQPNodeType::Aggregate) {
+      info.aggregate_nodes.emplace_back(std::static_pointer_cast<AggregateNode>(node), info.predicates.size());
+    } else if (node->type == LQPNodeType::Alias) {
+      info.alias_nodes.emplace_back(std::static_pointer_cast<AliasNode>(node));
+    } else if (node->type != LQPNodeType::Validate && node->type != LQPNodeType::Sort) {
+      // It is not safe to pull up predicates past this node, stop scanning
+      break;
+    }
+
+    DebugAssert(!node->right_input(), "Scan only implemented for nodes with one input");
```

Comment:
What "scan" do you refer to?

Reply:
This references an old name for this method (scan_for_join_predicates or something similar). I replaced it with a more sensible message